scrapy_po/webpage.py

```diff
@@ -66,6 +66,20 @@ class WebPage(PageObject, ResponseShortcutsMixin):
     response: ResponseData
 
 
+@attr.s(auto_attribs=True)
+class PageComponent(PageObject, ResponseShortcutsMixin):
+    """
+    Use it for implementing extractors for particular components of the
+    page (i.e. breadcrumbs) that can be reused across different page objects.
+
+    For example, a book page object can rely on breadcrumbs page component
+    to extract the book category and also a book listings page object could
+    rely in this component to extract the category a listings page
+    belongs to.
+    """
+    response: ResponseData
```

Comment:
I think it makes sense to have some base class which doesn't have ResponseShortcutsMixin or ResponseData to inherit from. With the current implementation it won't make sense to inherit AutoExtract-based page components from this PageComponent class.

Reply:
I see. Should we introduce two instead?:
```py
class PageComponent(PageObject):
   pass

@attr.s(auto_attribs=True)
class PageWebComponent(PageComponent, ResponseShortcutsMixin):
    response: ResponseData
```

Another option is to just provide a PageComponent extending from PageObject and leave  the user the responsibility to add the response and extend the mixin when they need web data:

```py
@attr.s(auto_attribs=True)
class MyBreadcrumbsComponent(PageComponent, ResponseShortcutsMixin):
    response: ResponseData
```

It doesn't look so easy. So I would go for the first option. @kmike What do you think?

Reply:
I'd try to find a name for a base class which says "handle me in the dependency injection!", then inherit PageObject and PageComponent from this class, or just drop them both (as they do nothing), if the name is good enough.

```py
class ItemPage(Component, ABC):
    @abstractmethod
    def to_item(self) -> Dict:
        return {}


@attr.s(auto_attribs=True)
class WebPage(ItemPage, ResponseShortcutsMixin):
    response: ResponseData 


class BasePaginator(Component, ABC):
    @abstractmethod
    def page_urls(self) -> List[str]:
        return []
    


```

Reply:
Well, what about `Injectable`. It seems very aligned with that it really means. Something injectable can be injected using dependency injection. 

And I would then remove `PageObject` and `PageComponent` at all. So I like your proposal but renaming `Component` by `Injectable`: 
```py
class ItemPage(Injectable, ABC):
    @abstractmethod
    def to_item(self) -> Dict:
        return {}


@attr.s(auto_attribs=True)
class WebPage(ItemPage, ResponseShortcutsMixin):
    response: ResponseData 


class BasePaginator(Injectable, ABC):
    @abstractmethod
    def page_urls(self) -> List[str]:
        return []
``` 
I see it used in othe libraries with the same meaning: https://angular.io/api/core/Injectable 

We could also rename `InjectPageObjectsMiddleware` to `InjectionMiddleware` so that everything gets more aligned. And then Page Objects is just a way of organizing the extraction code by declaring classes that are injectables and that get injected. 

How that is sound?

A possible alternative would be to mark classes as "injectable" using a decorator (`@Injectable`), but it would require to always use this decorator on the concrete classes. It seems better, in this case, to mark the pages by inheritance so that the "injectable" property is inherited, so I would discard it.

Reply:
Names.. On one hand, Injectable looks good: these are things which we can inject. But we have 2 types of DI going on:

1) Injecting these PageObjects and components in callbacks
2) Injecting ResponseData, etc. to PageObjects and components

Do we need to separate them, what do you think? I.e. should ResponseData be Injectable? It seems to go against the current design, but I haven't though that much about it.

Marking classes as injectable using decorator is not needed, because you can already mark them using Injectable.register(MyClass), if you don't want to use inheritance for some reason (maybe the class is out of your control, you don't want to change it? not sure what's the real use case). This is a feature of Python's abc module. Injectable.register can also be used as a class decorator - exactly what you proposed here.  See https://docs.python.org/3/library/abc.html#abc.ABCMeta.register

Reply:
I see two types of classes from the DI point of view:
1) Injectables. They are automatically built by the DI mechanism. Example: PageObjects
2) Providables. They are "external" classes that require a provider to be built, so they are not directly built by the DI, but instead rely on a  provider. Example: `ResponseData`, `AutoextractProductResponse`. 

Currently, it is not required to mark "providable" classes as "injectable", because an individual registry is kept. When the middleware is looking at which types can be provided, it is considering both types ("injectables" and "providables"). See https://github.com/scrapinghub/scrapy-po/blob/page_component/scrapy_po/middleware.py#L59-L66. (I know you already know, I'm writing it to sorting my mind)

Providable classes always take precedence over Injectable ones: if you found a provider for a class, you use it, even if the class is "Injectable". 

There is then no technical benefit in forcing the annotation of "providable" classes as "injectable" so the question is if "injectable" is a good name given that there are things (the "providables") that will be also be injected even without being "injectables". The two options I see to improve coherence are:
1) Force that all providable classes to be also Injectables (i.e. include a check that fails if they are not). 
2) Put a different name, like `BuildByInjectionButNotFromProvider` (obviously a horrendous name, but I can't think in something good meaning that). 

I don't like any of both options above. From my point of view, the ambiguity of `Injectable` is manageable, so I would stay with this proposal.  

Good point regarding that using `ABCMeta.register` can be used as a decorator. So we have then both options.

Reply:
I also dislike forcing ResponseData and alike objects being Injectable.

Ok, I'm fine with proceeding with Injectable. Still not a fan, but it is definitely better than what we have now, and we can change the name later if a better proposal arrives, it is early days.

Reply:
Another point against forcing ResponseData as being injectable: This objects that are "providable" can be optional (used in Unions), but that not's the case for the rest of injectable objects.

> Ok, I'm fine with proceeding with Injectable. Still not a fan

Yes, we can change it in the future. I'll change this PR to use `Injectable` by now.